Playwright: Part 2. Migrate rest of wp-likes spec and write additional cases.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR migrates rest of the [`wp-likes-spec.js`](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/specs-calypso/wp-likes-spec.js) file from the Selenium suite.

As part of the migration, the test cases have been split up where possible:
- `wp-likes__comments-spec.js` test the likes feature for comments only.
- `wp-like__posts-spec.js` test the likes feature for posts only.

The rubric of cases covered by this migration looks like the following:

Like and Unlike
| | posts | comments |
| --- | --- | --- |
| new | ✓ | ✓ |
| existing |✓ | ✓ |

To implement the Likes tests, additional supporting library files have also been migrated/created where applicable.

- `CommentComponent` represents the comment section of a post. 
- `PublishedPostPage` represents the post.



#### Testing instructions

CI
- ensure tests pass without any issues on TeamCity.

Local
- pull the branch and ensure they run with `mocha --config .mocharc_playwright.yml specs/specs-playwright/`.

Likes: Comment should look like this:
```
  [WPCOM] Likes (Comment): (desktop) @parallel
    Comment and like on an existing post
Logging in as e2eflowtestinggutenbergsimple
      ✓ Log in (3065ms)
      ✓ Visit site (1191ms)
      ✓ Click on first post (254ms)
      ✓ Post a comment (3712ms)
      ✓ Like a comment (1897ms)
    Comment and like on a new post
Logging in as e2eflowtestinggutenbergsimple
      ✓ Log in (2790ms)
      ✓ Start new post (5595ms)
      ✓ Enter post title (261ms)
      ✓ Enter post text (670ms)
      ✓ Publish and visit post (6067ms)
      ✓ Post a comment (3262ms)
      ✓ Like a comment (1994ms)
      ✓ Unlike a comment (616ms)


  13 passing (38s)
```

Likes: Post should look like this:
```
  [WPCOM] Likes (Post): (desktop) @parallel
    New post
Logging in as e2eflowtestinggutenbergsimple
      ✓ Log in (3007ms)
      ✓ Start new post (6651ms)
      ✓ Enter post title (330ms)
      ✓ Enter post text (676ms)
      ✓ Publish and visit post (5540ms)
      ✓ Like post (1080ms)
      ✓ Unlike post (96ms)
    Existing post
Logging in as e2eflowtestinggutenbergsimple
      ✓ Log in (2724ms)
      ✓ Visit site (849ms)
      ✓ Click on first post (321ms)
      ✓ Like post (1246ms)


  11 passing (34s)
```

Parent: #53169

worldomonation committed:
Migrate additional test cases from wp-like.

worldomonation committed:
Remove console.log and update JSDoc on CommentLikesComponent.
Update test.

worldomonation committed:
Restructure how variables are being shared in test.
Additional await selector safeguards.

worldomonation committed:
Separate commentLikesComponent so that the variable isn't being shared.

worldomonation committed:
Comments and annotations for tricky parts of the code.

worldomonation commented on packages/calypso-e2e/src/lib/components/post-likes-component.ts:
```diff
@@ -23,7 +23,7 @@ const selectors = {
  * @augments {BaseContainer}
  */
-export class LikesComponent extends BaseContainer {
+export class PostLikesComponent extends BaseContainer {
```
It did not make sense to try to unify the components for likes, since it would make the flow difficult to follow.

fullofcaffeine reviewed (approved):
Didn't do a very thorough review of the code as I'm still getting acquainted to it, but this tests well!

WunderBart commented on packages/calypso-e2e/src/lib/components/comments-component.ts:
```diff
@@ -0,0 +1,38 @@
+	 */
+	async postComment( comment: string ): Promise< void > {
+		await this.page.waitForSelector( selectors.commentTextArea );
+		await this.page.fill( selectors.commentTextArea, comment );
```
According to [PW docs](https://playwright.dev/docs/api/class-page#page-fill), the `page.fill` "waits for an element matching selector, waits for actionability checks, focuses the element, fills it and triggers an input event after filling". Do we still need to explicitly wait for the element before filling?

worldomonation replied:
I've typically found things to be more reliable when I explicitly wait for them in Calypso, especially around things that are in iframes (eg. like buttons) or fade in. I can probably remove this line for comments and it should work, but it's a general heads up.

worldomonation replied:
Removed extra wait and it appears to be passing locally and on TeamCity.

WunderBart commented on packages/calypso-e2e/src/lib/components/comments-likes-component.ts:
```diff
@@ -0,0 +1,81 @@
+	 * @returns {Promise<void>} No return value.
+	 */
+	async _postInit(): Promise< void > {
+		await this.page.waitForLoadState( 'networkidle' );
```
Would it make sense to always wait for `networkidle` in the `BaseContainer`?

worldomonation replied:
Not necessarily. Always waiting for `networkidle` would introduce a large amount of 'dead time' waiting around.

It does seem like every other component I throw in a `networkidle` but for many others where `domcontentloaded` would be more appropriate, using `networkidle` would be overkill.

Now, this mostly stems from the three level of control Playwright offers - `load`, `domcontentloded`, `networkidle`. The first two should handle most situations, but for things like `MyHomePage` I have found the missing middle between `domcontentloaded` and `networkidle` - perhaps something like `rendered`. So I err on the side of caution and wait for `networkidle` to be fired to ensure all requests have been processed.

WunderBart commented on packages/calypso-e2e/src/lib/components/comments-likes-component.ts:
```diff
@@ -0,0 +1,81 @@
+	 * @throws {Error} If no comments exist.
+	 */
+	async clickLikeComment( commentNum = 1 ): Promise< void > {
+		await this.page.waitForSelector( selectors.topLevelComments );
+
+		// For now, only works with top level comments since child comments introduce
+		// a whole new layer of complexity.
+		const topLevelComments = await this.page.$$( selectors.topLevelComments );
```
Would this work when replaced with the following?
```suggestion
		const topLevelComments = await this.page.waitForSelector( selectors.topLevelComments );
```

worldomonation replied:
Here, the `$$` selector returns `ElementHandle[]` which is what I want to do here.

`page.waitForSelector` matches on only one element and would return the first one that meets the criteria (visible, etc). Since I am trying to retrieve the entire list of top level comments for further processing, the `$$` call is more appropriate here.

The reason why I excluded child comments has to do with a few reasons:
- inconsistent selector - there isn't a good way to select `just` the child comments that I've found so far.
- what comment number to assign child comments - if the first comment has a reply, would that be considered second? What if there's further replies along? Would these comments bump rest of the comments down on the index?

On top of that, there isn't a test that tests replies/child comments at this moment in time. I figured I'd address the complexities of abstracting parent/child comments down the road when I write new tests.

WunderBart replied:
Ah, TIL that `page.$$` is basically a `document.querySelectorAll` 👍 

Regarding the post selector - maybe we could base the selector off the comment's text instead of number? We're posting the comment we're liking, so it should be all we need to know to make a selector:

```js
function like( comment ) {
  await this.page.click( `a.comment-like-link:below(p:text(${ comment }))` );
  // ...and await here for the comment to become liked
}
```

This approach also feels more user-oriented to me - we're (un)liking the comment based on the actual comment text.

worldomonation replied:
That's an idea, yeah. I can add a code path for liking a comment based on the comment content, not by the nth order by which it appears on page.

I still want to keep the code path for liking based on the nth number though, because I can see it being potentially useful for other tests.

worldomonation replied:
I've implemented your suggested changes in the latest push (support selection of comment by the comment body) while keeping the `nth` based selector path as well.

WunderBart commented on packages/calypso-e2e/src/lib/components/comments-likes-component.ts:
```diff
@@ -0,0 +1,81 @@
+		await Promise.all( [ likeButton.click(), likeButton.waitForElementState( 'enabled' ) ] );
+
+		// Check for intended outcome depending on the original liked state.
+		if ( isLiked.length ) {
```
nit: `isLiked` suggests the value is a bool. Can we make it i.e. `likedComments` (according to the used selector)?

worldomonation replied:
Originally I had it as `if (!!isLiked)` and wondered why for the longest time the corresponding click wasn't happening.

As I found out to my surprise `[]` was being evaluated as truthy. I came from python2/3 where that's falsy, so it really threw me for a loop.

Anyway, I can change that variable name to better reflect what it actually holds. Just figured I'll tell a story.

worldomonation replied:
I took another look, and the selector used is `selectors.liked` which maps to `.comment-liked`. This selector actually selects the presence of the `liked by ...` content, not the comment itself. 

I can probably change Line 68 to `await comment.isVisible( selectors.liked );` and that would return a boolean for us to work with.

But this is probably moot point seeing your later code review comment that rightfully point out, that `clickLike` on both comment and post is doing a lot of things behind the stage and it isn't necessarily transparent.

worldomonation replied:
I've refactored this portion significantly in [this commit](https://github.com/Automattic/wp-calypso/pull/53291/commits/2b3ed663a3845b87ef504de376044406576dd315). 

The changes there remove the need for `isLiked` as I have chosen to split the `clickXXX` methods into discrete `like` and `unlike` portions. All that's needed is after the click action takes place, the methods verify the appropriate selectors are visible on the DOM.

WunderBart commented on packages/calypso-e2e/src/lib/pages/my-home-page.ts:
```diff
@@ -11,10 +11,29 @@ import { Page } from 'playwright';
 const selectors = {
 	dashboard: '.customer-home__main',
 	statsCard: '.stats',
+	visitSiteButton: 'text=Visit site',
```
Same as https://github.com/Automattic/wp-calypso/pull/53323#discussion_r648149864. I can see there's an issue for it already, so that's awesome! :D https://github.com/Automattic/wp-calypso/issues/53556

worldomonation replied:
This isn't/wasn't in the scope of #53556, but do you think it should be?

This selector seems to be working fine as is, and while I don't discount that another `Visit site` text might show up later down the road - I don't think it would be attached to the DOM earlier than this specific button.

worldomonation replied:
I've added a narrowing selector in the most recent revision, it now reads `.customer-home__view-site-button >> text=Visit site`

WunderBart replied:
I might need to take back my suggestion here. Sorry for the confusion, I'm learning too! 😬  Just read the Good Practices for the PW Selectors section and it seems that avoiding non-user-facing attributes is actually [considered a good practice](https://playwright.dev/java/docs/selectors#prioritize-user-facing-attributes). Now that I think about it, it's been often an issue with Selenium errors that the test got broken because a class selector changed. Maybe PW has a more efficient mechanism for querying with text only and it doesn't really make sense to make those selectors more specific? What do you think - should we go back and drop the class selectors?

worldomonation replied:
Ah, I should have linked you that page! 

You are right that Selenium has been pretty brittle because of the selectors - mostly xpath, which I try to avoid, and CSS selectors. I just ran into this earlier as I was developing the `Media Edit` spec, since my personal WPCOM account seems to have a different image editor than what's available to the general public. If I had used user facing attributes most of this could have been avoided.

> Maybe PW has a more efficient mechanism for querying with text only and it doesn't really make sense to make those selectors more specific?

Playwright is indeed quite heavy on text selectors. For most things, text selectors alone function just fine, but as we've seen in the case of sidebars in #53566, sometimes the text selectors will match multiple things and the first match is not necessarily the one we want. Now, use of specific CSS selectors are often inevitable, but where possible we should try to use the simplest form possible.

Looping back to what you've said re: dropping class selectors - I think we should leave it is for now, as I do not anticipate the home dashboard changing anytime soon. We can revisit once tests begin failing. How is that?

WunderBart replied:
Sounds good. I guess it's about finding a way between sticking to good practices and making sure we're targeting the right element. A single classname shouldn't be that brittle 👍

WunderBart commented on packages/calypso-e2e/src/lib/pages/my-home-page.ts:
```diff
@@ -11,10 +11,29 @@ import { Page } from 'playwright';
+	 * @returns {Promise<void>} No return value.
+	 */
+	async visitSite(): Promise< void > {
+		await this.page.waitForSelector( selectors.visitSiteButton );
+		await Promise.all( [
+			this.page.waitForNavigation(),
+			this.page.click( selectors.visitSiteButton ),
```
If `page.click` waits for a selector on its own, do we still need to explicitly wait for it?
```suggestion
		await Promise.all( [
			this.page.waitForNavigation(),
			this.page.click( selectors.visitSiteButton ),
```

worldomonation replied:
Similar to the comment left earlier - it may be prudent, but perhaps not strictly necessary.

worldomonation replied:
It was not necessary and so I have removed the wait.

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes__comment-spec.js:
```diff
@@ -0,0 +1,95 @@
+
+		it( 'Visit site', async function () {
+			const myHomePage = await MyHomePage.Expect( this.page );
+			await myHomePage.visitSite();
```
There's the [`Visit` static method in the ole `AsyncBaseContainer`](https://github.com/Automattic/wp-calypso/blob/e41c8158b83e89036cd1c6ccd313d7ea02ab17e0/test/e2e/lib/async-base-container.js#L34-L42). Do you think we should also have one for the new `BaseContainer`?

worldomonation replied:
I was not a fan of the `Visit` static method, but can't quite communicate why I feel that way. 

It's something I am open to implementing. Will file a task on the board.

worldomonation replied:
Task filed as #53601 :)

WunderBart commented on packages/calypso-e2e/src/lib/components/comments-likes-component.ts:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @augments {BaseContainer}
+ */
+export class CommentsLikesComponent extends BaseContainer {
```
Would it make sense to merge it with the `CommentsComponent`?

worldomonation replied:
I can look into that for this PR. Do you think it would make sense to also rename the `PostLikesComponent` into a `PostPage`, analogous to the change proposed for comments?

worldomonation replied:
Edit: I've removed the nested class. It felt like a good idea, but upon closer inspection it was redundant - it was simply better to incorporate the methods into the PublishedPostsPage itself.

~~I have made changes using an approach that I hope is not too controversial: inner/nested class.~~

~~Previously (as you have noticed) there were arguably too many components/pages at play:~~

- PostLikesComponent to interact with the post like button.
- CommentComponent to interact with the comment field.
- CommentsLikesComponent to interact with the comment like button.
- PublishedPostsListPage to represent the page on the published site that lists all posts.

~~It occurred to me that there was no page object to represent a single published post. In my mind, a revised hierarchy should look like this:~~

```
          PublishedPostsListPage
                   v
          PublishedPostsPage
            / (has)        \ (has)
PostLikesComponent       CommentsComponent

```

~~The inner class was my attempt at codifying the `has` relationship, but I am on the fence as to whether this is a good idea.~~

~~I'd like to hear what you think. Do you think it may be anti-pattern to use inner classes?~~

WunderBart commented on test/e2e/specs/specs-playwright/wp-likes__post-spec.js:
```diff
@@ -0,0 +1,82 @@
+		} );
+
+		it( 'Unlike post', async function () {
+			await postLikesComponent.clickLikePost();
```
I think it would be better to separate the like & unlike helpers. Currently, it works as a toggle, which makes the `unlike` rely on a valid outcome of the `like` step. In theory, the `like` step could be actually unliking the post and the `unlike` step would be doing the opposite, and we wouldn't know that it happened 😄

worldomonation replied:
Guess I went too far in trying to simplify the test writing experience for test writers eh?
I was hoping the `clickLikePost` would be an one-stop shop to perform all actions and verification on the like state. You have a good point in that it doesn't explicitly communicate what action was taken.

I guess time to split up this method into two actions.

worldomonation replied:
As mentioned in [this comment](https://github.com/Automattic/wp-calypso/pull/53291#discussion_r650145720) I have split up the methods into discrete actions.

worldomonation committed:
Address smaller issues on the PR.

worldomonation committed:
Split the clickLike functionality into two parts - like and unlike - for both Comments and Posts.

Create a new page, PublishedPostPage to represent a particular published post's page.
Merge PostLikesComponent into the PublishedPostPage as its private inner class.
Expose functionality of the inner PostLikesComponent class through functions on the outer class.

Merge CommentsComponent and CommentsLikesComponent into one.

worldomonation committed:
Another refactor to remove the inner class from PublishedPostPage.

worldomonation committed:
Formatting and JSDoc fixes.

worldomonation committed:
Refactor how the like/unlike comment process occurs.
- discontinue use of `this.page.$$` when a numerical selector is provided. Instead, select the comment using `nth-match`.
- support matching the comment by its text body.

worldomonation committed:
Fix how selectors were being passed into the like and unlike methods.
Remove errant page.pause().

worldomonation commented:
Oh shoot, I accidentally included the test logs.

worldomonation committed:
Remove accidental inclusion of junit xml files.
